polygon: race in bridge

```
==================
WARNING: DATA RACE
Write at 0x00c001246018 by goroutine 26:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Run()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:120 +0x58c
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge.func1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:109 +0xa0
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge.gowrap1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:117 +0x54

Previous read at 0x00c001246018 by goroutine 25:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Synchronize()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:188 +0x5c
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:119 +0xb54
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 26 (running) created at:
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:106 +0xaa0
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 25 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x5e4
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2161 +0x80
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2159 +0x6e0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2027 +0xb74
  main.main()
      _testmain.go:51 +0x294
==================
--- FAIL: TestBridge (0.06s)
    database.go:51: INFO[08-02|11:49:14.448] Opening Database                         label=polygon-bridge path=/var/folders/f4/htghck_11274s_4nwxnmjsn40000gn/T/TestBridge1615623039/001/polygon-bridge
    kv_mdbx.go:380: DBUG[08-02|11:49:14.481] [db] open                                label=polygon-bridge sizeLimit=16GB pageSize=16384
    bridge.go:96: DBUG[08-02|11:49:14.486] [bridge] Bridge is running               lastEventID=0
    bridge.go:126: DBUG[08-02|11:49:14.490] [bridge] got 3 new events, last event ID: 3, ready: false 
    bridge.go:163: DBUG[08-02|11:49:14.491] [bridge] Creating map for block 4, start ID 0, end ID 2 
    bridge.go:222: DBUG[08-02|11:49:14.491] [bridge] got 2 events for block 4 
    bridge.go:222: DBUG[08-02|11:49:14.491] [bridge] got 2 events for block 4 
    testing.go:1398: race detected during execution of test
==================
WARNING: DATA RACE
Write at 0x00c001246798 by goroutine 28:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Run()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:113 +0x4bc
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind.func1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:213 +0xa0
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind.gowrap1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:221 +0x54

Previous read at 0x00c001246798 by goroutine 27:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Synchronize()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:188 +0x5c
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:223 +0xcbc
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 28 (running) created at:
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:210 +0xc08
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 27 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x5e4
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2161 +0x80
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2159 +0x6e0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2027 +0xb74
  main.main()
      _testmain.go:51 +0x294
==================
==================
WARNING: DATA RACE
Write at 0x00c001246798 by goroutine 28:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Run()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:120 +0x58c
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind.func1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:213 +0xa0
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind.gowrap1()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:221 +0x54

Previous read at 0x00c001246798 by goroutine 27:
  github.com/erigontech/erigon/polygon/bridge.(*Bridge).Synchronize()
      /Users/shohamc1/erigon/polygon/bridge/bridge.go:188 +0x5c
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:223 +0xcbc
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 28 (running) created at:
  github.com/erigontech/erigon/polygon/bridge_test.TestBridge_Unwind()
      /Users/shohamc1/erigon/polygon/bridge/bridge_test.go:210 +0xc08
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x40

Goroutine 27 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1742 +0x5e4
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2161 +0x80
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1689 +0x180
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2159 +0x6e0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2027 +0xb74
  main.main()
      _testmain.go:51 +0x294
==================
```

--- a/polygon/bor/bor.go
+++ b/polygon/bor/bor.go
@@ -39,6 +39,7 @@ import (
 	"golang.org/x/sync/errgroup"
 
 	"github.com/erigontech/erigon-lib/common/dbg"
+	"github.com/erigontech/erigon/polygon/bridge"
 
 	"github.com/erigontech/erigon-lib/log/v3"
 
@@ -1511,6 +1512,10 @@ func (c *Bor) CommitStates(
 	if c.bridgeReader != nil {
 		events, err := c.bridgeReader.Events(c.execCtx, blockNum)
 		if err != nil {
+			if errors.Is(err, bridge.ErrMapNotAvailable) {
+				return nil
+			}
+
 			return err
 		}
 

--- a/polygon/bridge/bridge.go
+++ b/polygon/bridge/bridge.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"sync/atomic"
 	"time"
 
 	"github.com/erigontech/erigon-lib/kv"
@@ -42,9 +43,9 @@ type fetchSyncEventsType func(ctx context.Context, fromId uint64, to time.Time, 
 
 type Bridge struct {
 	store                    Store
-	ready                    bool
-	lastProcessedBlockNumber uint64
-	lastProcessedEventID     uint64
+	ready                    atomic.Bool
+	lastProcessedBlockNumber atomic.Uint64
+	lastProcessedEventID     atomic.Uint64
 
 	log                log.Logger
 	borConfig          *borcfg.BorConfig
@@ -61,14 +62,12 @@ func Assemble(dataDir string, logger log.Logger, borConfig *borcfg.BorConfig, fe
 
 func NewBridge(store Store, logger log.Logger, borConfig *borcfg.BorConfig, fetchSyncEvents fetchSyncEventsType, stateReceiverABI abi.ABI) *Bridge {
 	return &Bridge{
-		store:                    store,
-		log:                      logger,
-		borConfig:                borConfig,
-		fetchSyncEvents:          fetchSyncEvents,
-		lastProcessedBlockNumber: 0,
-		lastProcessedEventID:     0,
-		stateReceiverABI:         stateReceiverABI,
-		stateClientAddress:       libcommon.HexToAddress(borConfig.StateReceiverContract),
+		store:              store,
+		log:                logger,
+		borConfig:          borConfig,
+		fetchSyncEvents:    fetchSyncEvents,
+		stateReceiverABI:   stateReceiverABI,
+		stateClientAddress: libcommon.HexToAddress(borConfig.StateReceiverContract),
 	}
 }
 
@@ -90,7 +89,7 @@ func (b *Bridge) Run(ctx context.Context) error {
 		return err
 	}
 
-	b.lastProcessedEventID = lastProcessedEventID
+	b.lastProcessedEventID.Store(lastProcessedEventID)
 
 	// start syncing
 	b.log.Debug(bridgeLogPrefix("Bridge is running"), "lastEventID", lastEventID)
@@ -110,20 +109,20 @@ func (b *Bridge) Run(ctx context.Context) error {
 		}
 
 		if len(events) != 0 {
-			b.ready = false
+			b.ready.Store(false)
 			if err := b.store.AddEvents(ctx, events, b.stateReceiverABI); err != nil {
 				return err
 			}
 
 			lastEventID = events[len(events)-1].ID
 		} else {
-			b.ready = true
+			b.ready.Store(true)
 			if err := libcommon.Sleep(ctx, 30*time.Second); err != nil {
 				return err
 			}
 		}
 
-		b.log.Debug(bridgeLogPrefix(fmt.Sprintf("got %v new events, last event ID: %v, ready: %v", len(events), lastEventID, b.ready)))
+		b.log.Debug(bridgeLogPrefix(fmt.Sprintf("got %v new events, last event ID: %v, ready: %v", len(events), lastEventID, b.ready.Load())))
 	}
 }
 
@@ -154,19 +153,19 @@ func (b *Bridge) ProcessNewBlocks(ctx context.Context, blocks []*types.Block) er
 
 		prevSprintTime = time.Unix(int64(block.Time()), 0)
 
-		lastDBID, err := b.store.GetSprintLastEventID(ctx, b.lastProcessedEventID, timeLimit, b.stateReceiverABI)
+		lastDBID, err := b.store.GetSprintLastEventID(ctx, b.lastProcessedEventID.Load(), timeLimit, b.stateReceiverABI)
 		if err != nil {
 			return err
 		}
 
-		if lastDBID > b.lastProcessedEventID {
-			b.log.Debug(bridgeLogPrefix(fmt.Sprintf("Creating map for block %d, start ID %d, end ID %d", block.NumberU64(), b.lastProcessedEventID, lastDBID)))
-			eventMap[block.NumberU64()] = b.lastProcessedEventID
+		if lastDBID > b.lastProcessedEventID.Load() {
+			b.log.Debug(bridgeLogPrefix(fmt.Sprintf("Creating map for block %d, start ID %d, end ID %d", block.NumberU64(), b.lastProcessedEventID.Load(), lastDBID)))
+			eventMap[block.NumberU64()] = b.lastProcessedEventID.Load()
 
-			b.lastProcessedEventID = lastDBID
+			b.lastProcessedEventID.Store(lastDBID)
 		}
 
-		b.lastProcessedBlockNumber = block.NumberU64()
+		b.lastProcessedBlockNumber.Store(block.NumberU64())
 	}
 
 	err := b.store.StoreEventID(ctx, eventMap)
@@ -185,7 +184,7 @@ func (b *Bridge) Synchronize(ctx context.Context, tip *types.Header) error {
 			return ctx.Err()
 		default:
 		}
-		if b.ready || b.lastProcessedBlockNumber >= tip.Number.Uint64() {
+		if b.ready.Load() || b.lastProcessedBlockNumber.Load() >= tip.Number.Uint64() {
 			return nil
 		}
 	}
@@ -208,7 +207,7 @@ func (b *Bridge) Events(ctx context.Context, blockNum uint64) ([]*types.Message,
 	}
 
 	if end == 0 { // exception for tip processing
-		end = b.lastProcessedEventID
+		end = b.lastProcessedEventID.Load()
 	}
 
 	eventsRaw := make([]*types.Message, 0, end-start+1)

--- a/polygon/bridge/bridge.go
+++ b/polygon/bridge/bridge.go
@@ -199,10 +199,6 @@ func (b *Bridge) Unwind(ctx context.Context, tip *types.Header) error {
 func (b *Bridge) Events(ctx context.Context, blockNum uint64) ([]*types.Message, error) {
 	start, end, err := b.store.GetEventIDRange(ctx, blockNum)
 	if err != nil {
-		if errors.Is(err, ErrMapNotAvailable) {
-			return nil, nil
-		}
-
 		return nil, err
 	}
 

--- a/polygon/bridge/db.go
+++ b/polygon/bridge/db.go
@@ -20,8 +20,6 @@ import (
 	"bytes"
 	"context"
 	"encoding/binary"
-	"errors"
-	"fmt"
 	"time"
 
 	"github.com/erigontech/erigon-lib/kv"
@@ -292,7 +290,7 @@ func (s *MdbxStore) GetEventIDRange(ctx context.Context, blockNum uint64) (uint6
 		return start, end, err
 	}
 	if v == nil { // we don't have a map
-		return start, end, errors.New(fmt.Sprintf("map not available for block %d", blockNum))
+		return start, end, ErrMapNotAvailable
 	}
 
 	err = binary.Read(bytes.NewReader(v), binary.BigEndian, &start)

